ratelimit/keeper/genesis.go

```diff
@@ -35,17 +39,41 @@ func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState)
 		}
 		k.SetPendingSendPacket(ctx, channelId, sequence)
 	}
+
+	// Verify the epoch hour duration is specified
+	if genState.HourEpoch.Duration == 0 {
+		panic("Hour epoch duration must be specified")
+	}
+
+	// If the hour epoch has been initialized already (epoch number != 0), validate and then use it
```

Comment:
This would be in the case where you're not starting from true "genesis", but were creating a fork of the chain starting from a previous export

It's possible my mental model of these genesis functions is flawed, but my prior was that they should be built in a way that can handle starting from block 0, but also starting from some random state as some arbitrary point in time

Reply:
Yeah, if you export the state and then you start the chain from that state, you get into the first if condition (`EpochNumber > 0`) and indeed you need to just write the epoch info into the store (i.e., `SetHourEpoch`). For me it's a bit more confusing the other part, when you add the ratelimit module for the first time. In this case `InitGenesis` is called, but I'm not sure where the `GenesisState` comes from. Is it the `DefaultGenesis`? If so, then indeed the `EpochNumber` is set to `0` and the `Duration` to one hour. In this case, indeed we need to initialize the HourEpoch struct and the code below looks good. Maybe a nit, why not initialize the EpochNumber to 1 instead of a number in `[0, 23]`?

Reply:
yeah that summary is right - default genesis has epoch number 0 and duration 1, and then the actual epoch gets initialized in InitGenesis

Reason for the epoch number being non-1 is mostly a nit UX improvement. I think the most common rate limit window is 24 hours which is reset when `HourEpochNumber % 24 == 0`. So by setting the epoch number to the current hour, the rate limits will always reset at 0:00 UTC, regardless of when the upgrade took place. Otherwise, they would reset at the hour closest to the upgrade time